agent/engine/common_windows_integ_testutil.go

```diff
@@ -45,7 +45,7 @@ func GetLongRunningCommand() []string {
 	return []string{"ping", "-t", "localhost"}
 }
 
-// TODO: implement this
 func isDockerRunning() bool {
-	return true
+	_, err := os.Stat("//./pipe/docker_engine")
```

Comment:
I don't think the os.Stat implementation will work.  In my testing it always returned the same result regardless of whether docker is running or not.
Here's an alternative implementation that does work.

```
package main

import (
    "flag"
    "golang.org/x/sys/windows/svc/mgr"
    "log"
)

func isDockerRunning() bool {
    var svcName = "docker"
    flag.StringVar(&svcName, "name", svcName, "name of the service")
    flag.Parse()
    svcMgmt, err := mgr.Connect()
    if err != nil {
        log.Fatalf("Error connecting to service manager: %v", err)
    }

    dockerSvc, err := svcMgmt.OpenService(svcName)
    if err != nil {
        log.Fatalf("Error opening service %q: %v", svcName, err)
    }

    dockerStatus, err := dockerSvc.Query()
    if err != nil {
        log.Fatalf("Error querying docker service: %v", err)
    }
    log.Printf("State:%v, ProcessId:%v, Win32ExitCode: %v", dockerStatus.State, dockerStatus.ProcessId, dockerStatus.Win32ExitCode)
    return dockerStatus.Win32ExitCode == 0
}

func main() {
    dockerStatus := isDockerRunning()
    if dockerStatus != true {
        log.Fatalf("Docker is not running")
    }
    log.Printf("Docker is running")
}
```
```
2024/08/30 13:55:25 State:4, ProcessId:4500, Win32ExitCode: 0
2024/08/30 13:55:25 Docker is running
```
```
2024/08/30 13:55:04 State:1, ProcessId:0, Win32ExitCode: 1067
2024/08/30 13:55:04 Docker is not running
```

Reply:
I have minimal knowledge/experience with Windows development, so bear with me while I explain my testing process...

To exercise the `os.Stat` implementation of `isDockerRunning`, I added a test
```
func TestIsDockerRunning(t *testing.T) {
	assert.True(t, isDockerRunning())
}
```
which I ran with
```
go test -tags integration  -count 1 -v ./agent/engine/... -run "TestIsDockerRunning"
```
I ran `get-service docker` to check if docker was already running. It was, and the test passed.
I ran `stop-service docker` to stop docker. The test failed when I reran it, as was expected.
I ran `start-service docker` to restart docker. The test passed.

Am I not starting/stopping docker correctly? I got the same results when I swapped out the `os.Stat` implementation with the service manager implementation.

Reply:
Based on your test results, it seems that Go is handling it properly if run as part of the integration test framework.  Given your test results, I'm OK with your earlier implementation since it is very simple and it works.  (Plus, the implementation I provided should be modified to include deferred closures to the service manager and service instance handle.)

My concern was whether Go's `os.Stat` method can differentiate between named pipes and regular on-disk files.  Named pipes are accessed via file system drivers (NPFS), but they are are not part of the actual on-disk file system, which is (typically) NTFS.

Running the code snippet, as-is, on Windows in a stand-alone way doesn't seem to work right, however.  That said, connecting through service manager requires the unit tests to be run as an administrator.  


---

> Am I not starting/stopping docker correctly?

There is more than one way to start/stop services on Windows.  start-service docker and stop-service docker is the PowerShell way to start/stop docker and is perfectly acceptable.